### PR TITLE
feat: add identity pool and iam roles

### PR DESCRIPTION
**Pull request by adamelmore**
This PR adds richer IAM auth support by automating the creation of an Identity Pool and IAM roles based on the `@auth` directive transformations.

@kcwinner let me know your thoughts here. I've been manually keeping these in sync in my projects, and thought it would be nice to handle it based on the transformer output. If you'd rather go about this a different way, I'll take no offense!

If this seems like a good addition, I'll also update the readme to clarify that this support has been added.

One outstanding question of mine is whether we should automatically update the AppSync auth configuration to include IAM as an auth type if these roles are present in the transformer output. As it stands, we could generate these roles and an Identity Pool and they wouldn't be usable if the auth config passed in by the user didn't allow for IAM auth.

Lastly, I also need to add the Identity Pool to the resulting output so that users can interact with it (and get the pertinent details out of the stack). I wanted to get this in front of you sooner than later, and then make those changes if you're interested in this feature.

### CODE REVIEW

**Comment by adamelmore** [comment]
I should note that my usage of IAM auth is mostly using `public` to allow for some types on my schema to be publicly accessible in my apps (typically read-only). So, the unauthenticated role is helpful for that case. That said, I'm not entirely sure what the authenticated role utility would be.

My understanding is that by defining a directive like `@auth(rules: [{ allow: private, provider: iam }])`, you're giving that authenticated role access, but how do people typically authenticate against the identity pool? I just don't know enough about those use cases to understand the value of setting up this role. When I use `allow: private, provider: iam` I'm typically doing so for backend automation, where Lambdas are reacting to some event and calling my API to make changes that can be visible to users (through subscriptions). In those cases, I set up the permissions on my Lambda execution role for calling whichever GraphQL operations necessary, and I certainly wouldn't be authenticating against the Identity Pool.

I know that Cognito User Pools can serve as the identity provider for an Identity Pool, but if that were the usage in the AppSync scenario, I don't know why you wouldn't just use User Pool auth?

If you have a better understanding of this use case, I'd love to hear your thoughts.

**Comment by kcwinner** [comment]
I think there are two separate use cases for auth role, and this library isn't really trying to solve the latter (right now). In Amplify you can get an unauth role to perform actions on an AppSync API or with other AWS services. Or, you can get an auth role to perform IAM actions like upload S3 files, etc. The auth role for this transformer is unnecessary, in my opinion, because as you said - IAM based actions should be performed and scoped by your other resources (Lambda function, EC2 instance, etc). We aren't really worried about the "user" authorized role because to communicate with AppSync they just use the cognito token. I made sure to note that in the [README](https://github.com/kcwinner/cdk-appsync-transformer#iam). However, it may be a helpful convenience to support it.

As for the unauth role, I'm thinking we let the user pass in the Identity Pool and role(s) instead of creating them for them. I don't want to assume anything about the users stacks and setup. What if they already have an external pool? I think that's in the spirit of what I was trying to accomplish by being less opinionated than Amplify.

Basically, keep the portions you added for getting the resources from the role and applying them, but not actually creating them.

Now, we *could* offer to create a userpool or identitypool for them with other methods

**Comment by adamelmore** [comment]
Makes perfect sense! I'll make those changes over the weekend (or early next week).

**Comment by adamelmore** [comment]
Revisiting this, it occurs to me that if they were to pass in the auth and/or unauth roles, we could populate them with the policy statements derived from the `@auth` directives, but I don't think we'd do anything with an Identity Pool. Does it make sense to just have them optionally pass in the roles or am I missing something with the pool?

**Comment by adamelmore** [comment]
Another option would just be to expose the policy statements on the transformer such that users could grab them and add them to their roles as they wish. With that scenario, they're' not passing anything in. Let me know your thoughts!

**Comment by kcwinner** [comment]
> Revisiting this, it occurs to me that if they were to pass in the auth and/or unauth roles, we could populate them with the policy statements derived from the `@auth` directives, but I don't think we'd do anything with an Identity Pool. Does it make sense to just have them optionally pass in the roles or am I missing something with the pool?

This is what I was originally thinking. BUT now that you mention the other option of just exposing the policy statements I'm kind of liking that. Let me think on this.

Additionally, we could do both. If you pass in the roles we attach. Or, if you don't want to, expose them for later usage.

**Comment by adamelmore** [comment]
Just let me know what direction you want to take it; no rush!

**Comment by kcwinner** [comment]
@adamelmore how about we go with something that follows the IGrantable model?
Similar to what we talked about but following the same model as other L2 constructs

[CDK Docs](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-iam.IGrantable.html)

[DynamoDB grant](https://github.com/aws/aws-cdk/blob/46a16312825cc7b5bf985bc8b69146a1397690d0/packages/%40aws-cdk/aws-dynamodb/lib/table.ts#L347)

Example might be: 
```
const unauthRole = new Role(...);
const appsyncTransformer = new AppSyncTransformer(...);

appsyncTransformer.grantUnauth(unauthRole)
```

Or does this seem wrong? I can also take a swing at this if you want

**Comment by adamelmore** [comment]
Ah, this makes perfect sense, IMO.

I won't be able to jump back to this until middle of next week, so if you have a chance before that go for it!

**Comment by kcwinner** [comment]
Superseded by https://github.com/kcwinner/cdk-appsync-transformer/pull/204
